performance_manager/lib/postgres_utils.py

```diff
@@ -76,22 +79,114 @@ def get_aws_engine() -> sa.future.engine.Engine:  # type: ignore
-        logging.error("Error Writing Dataframe to Database")
-        logging.exception(e)
+
+    def __init__(self, experimental: bool = False, verbose: bool = False):
+        """
+        initialize db manager object, creates engine and sessionmaker
+        """
+        if experimental:
+            self.engine = get_experimental_engine(echo=verbose)
+
+            @sa.event.listens_for(sa.engine.Engine, "connect")
+            def set_sqlite_pragma(dbapi_connection, _):  # type: ignore
+                cursor = dbapi_connection.cursor()
+                cursor.execute("PRAGMA foreign_keys=ON")
+                cursor.close()
+
+        else:
+            self.engine = get_local_engine(echo=verbose)
+
+        self.session = sessionmaker(bind=self.engine)
+
+        # create tables in SqlBase
+        SqlBase.metadata.create_all(self.engine)
+
+    def _get_schema_table(self, table: Any) -> sa.sql.schema.Table:
+        if isinstance(table, sa.sql.schema.Table):
+            return table
+        if isinstance(table, sa.orm.decl_api.DeclarativeMeta):
+            # mypy error: "DeclarativeMeta" has no attribute "__table__"
+            return table.__table__  # type: ignore
+
+        raise TypeError(f"can not pull schema table from {type(table)} type")
+
+    def get_session(self) -> sessionmaker:
+        """
+        get db session for performing actions
+        """
+        return self.session
+
+    def execute(self, statement: Any) -> None:
+        """
+        execute db action without data
+        """
+        with self.session.begin() as cursor:
+            cursor.execute(statement)
+
+    def insert_dataframe(
+        self, dataframe: pandas.DataFrame, insert_table: Any
+    ) -> None:
+        """
+        insert data into db table from pandas dataframe
+        """
+        insert_as = self._get_schema_table(insert_table)
+
+        with self.session.begin() as cursor:
+            cursor.execute(
+                sa.insert(insert_as),
+                dataframe.to_dict(orient="records"),
+            )
+
+    def select_as_dataframe(
+        self, select_query: sa.sql.selectable.Select
+    ) -> pandas.DataFrame:
+        """
+        select data from db table and return pandas dataframe
+        """
+        with self.session.begin() as cursor:
+            return pandas.DataFrame(
+                [row._asdict() for row in cursor.execute(select_query)]
+            )
+
+    def select_as_list(
+        self, select_query: sa.sql.selectable.Select
+    ) -> List[Any]:
+        """
+        select data from db table and return list
+        """
+        with self.session.begin() as cursor:
+            return [row._asdict() for row in cursor.execute(select_query)]
+
+    def truncate_table(self, table_to_truncate: Any) -> None:
+        """
+        truncate db table
+        """
+        truncat_as = self._get_schema_table(table_to_truncate)
+
+        truncat_as.drop(self.engine)
+        truncat_as.create(self.engine)
```

Comment:
is this truncate or completely erase and stand back up?

Reply:
sqlalchemy doesn't have a strict truncate, so this totally drops a table from the database and then re-creates it, which is basically what a truncate does anyways